internal/test/e2e/vault_test.go

```diff
@@ -0,0 +1,218 @@
+//
+// After the configuration is synced in the DP, an HTTPS client is
+// created using the `goodCACertPEM` used to sign the
+// deployed certificate, and then a GET is performed to test the
+// proxy functionality, which should return a 200.
+func TestSecretsManagementEnvVault(t *testing.T) {
+	// ensure that Vaults can be synced to Kong gateway
+	cleanup := run.Koko(t)
+	defer cleanup()
+
+	c := httpexpect.New(t, "http://localhost:3000")
+
+	// create vault entity
+	vault := &v1.Vault{
+		Id:     uuid.NewString(),
+		Name:   "env",
+		Prefix: "my-env-vault",
+		Config: &v1.Vault_Config{
+			Prefix: "KONG_MY_SECRET_",
+		},
+	}
+
+	c.POST("/v1/vaults").WithJSON(vault).Expect().Status(http.StatusCreated)
+
+	// create service
+	service := &v1.Service{
+		Id:   uuid.NewString(),
+		Name: "foo",
+		Host: "mockbin.org",
+		Path: "/status/200",
+	}
+	c.POST("/v1/services").WithJSON(service).Expect().Status(http.StatusCreated)
+
+	// create route
+	route := &v1.Route{
+		Name:  "bar",
+		Paths: []string{"/r1"},
+		Service: &v1.Service{
+			Id: service.Id,
+		},
+	}
+	c.POST("/v1/routes").WithJSON(route).Expect().Status(http.StatusCreated)
+
+	// create certificate
+	certificate := &v1.Certificate{
+		Id:   uuid.NewString(),
+		Cert: "{vault://my-env-vault/cert}",
+		Key:  "{vault://my-env-vault/key}",
+	}
+	c.POST("/v1/certificates").WithJSON(certificate).Expect().Status(http.StatusCreated)
+
+	// create sni
+	sni := &v1.SNI{
+		Id:   uuid.NewString(),
+		Name: "localhost",
+		Certificate: &v1.Certificate{
+			Id: certificate.Id,
+		},
+	}
+	c.POST("/v1/snis").WithJSON(sni).Expect().Status(http.StatusCreated)
+
+	dpCleanup := run.KongDP(kong.GetKongConfForSharedWithSecrets())
+	defer dpCleanup()
+
+	require.Nil(t, util.WaitForKongAdminAPI(t))
+	kongClient.RunWhenKong(t, ">= 3.0.0")
+
+	expectedConfig := &v1.TestingConfig{
+		Vaults:       []*v1.Vault{vault},
+		Services:     []*v1.Service{service},
+		Routes:       []*v1.Route{route},
+		Certificates: []*v1.Certificate{certificate},
+		Snis:         []*v1.SNI{sni},
+	}
+	util.WaitFunc(t, func() error {
+		err := util.EnsureConfig(expectedConfig)
+		t.Log("configuration mismatch for vault", err)
+		return err
+	})
+
+	require.NoError(t, backoff.RetryNotify(func() error {
+		// build simple http client
+		client := &http.Client{}
+		// use simple http client with https should result
+		// in a failure due to missing certificate.
+		_, err := client.Get("https://localhost:8443/r1") //nolint:bodyclose
+		require.NotNil(t, err)
+
+		// use transport with wrong CA cert should result
+		// in a failure due to unknown authority.
+		badCACertPool := x509.NewCertPool()
+		badCACertPool.AppendCertsFromPEM(badCACertPEM)
+
+		client = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					RootCAs:    badCACertPool,
+					ClientAuth: tls.RequireAndVerifyClientCert,
+				},
+			},
+		}
+
+		_, err = client.Get("https://localhost:8443/r1") //nolint:bodyclose
+		require.NotNil(t, err)
```

Comment:
This is actually making sure we do have an error here.

Reply:
Ah, sorry, bad copy & pasta, meant to say `require.Error()` or `require.EqualError()`.

Reply:
Yeah I initially wanted to use `EqualError` here too, I tried this in the decK repo some time ago and found that the returned error is not consistent based on different Kong versions, so this may become quite flaky.

If it's not too bad, I'd leave it as it is.

Reply:
Ah, why not use `require.Error()` then? That only asserts that there was an error.

Reply:
Oh sure!